File: backend/app/schemas/product.py
from pydantic import BaseModel, Field

from app.schemas.category import CategorySummary
from app.schemas.product_image import ProductImageResponse

class ProductCreate(BaseModel):
    name: str
    category_id: int | None = None
    description: str | None = None
    price: float
    condition: str
    location: str | None = None
    contact_phone: str | None = None


class ProductUpdate(BaseModel):
    name: str
    category_id: int | None = None
    description: str | None = None
    price: float
    condition: str
    location: str | None = None
    contact_phone: str | None = None
    status: str


class ProductResponse(ProductCreate):
    id: int
    slug: str
    status: str
    category: CategorySummary | None = None
    images: list[ProductImageResponse] = Field(
        default_factory=list
    )

    model_config = {
        "from_attributes": True
    }

File: frontend/src/types/product.ts

export interface ProductImage {
  id: number
  storage_path: string
  image_url: string
  display_order: number
  is_primary: boolean
  created_at: string
}


export interface ProductCategory {
  id: number
  name: string
  slug: string
}


export interface Product {
  id: number
  name: string
  slug: string

  category_id: number | null
  category: ProductCategory | null

  description: string | null
  price: number
  condition: string

  location: string | null
  contact_phone: string | null

  status: string

  images: ProductImage[]
}


export interface ProductCreate {
  name: string
  category_id?: number | null
  description?: string | null
  price: number
  condition: string
  location?: string | null
  contact_phone?: string | null
}


export interface ProductUpdate {
  name: string
  category_id?: number | null
  description?: string | null
  price: number
  condition: string
  location?: string | null
  contact_phone?: string | null
  status: string
}

export type ImageBackground =
  | 'white'
  | 'grey'
  | 'black'


export type ImageRotation =
  | 0
  | 90
  | 180
  | 270


export interface ImageCrop {
  x: number
  y: number
  width: number
  height: number
}


export interface ImageEditRequest {
  background: ImageBackground
  rotation: ImageRotation
  crop: ImageCrop | null
}


export interface ImageProcessResponse {
  image_base64: string
  content_type: string
}